### PR TITLE
Add ActiveMQ based MB support for cache invalidation

### PR DESCRIPTION
**Pull request by inthirakumaaran**
## Related Issue
https://github.com/wso2/product-is/issues/4355

### CODE REVIEW

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/ActiveMQProducer.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                log.info("......................sending from producer...................." +
+                        clusterCacheInvalidationRequest.toString());
+
+                // Clean up resources
+                producer.close();
+                session.close();
+                connection.close();
+                break;
+            } catch (JMSException e) {
+//                throw new CacheException("Something went wrong with activeMQ producer", e);
```
Comment (inthirakumaaran) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/ActiveMQProducer.java]: Need to remove this exception

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/CacheSyncUtils.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    public static final String ACTIVEMQ_PRODUCER_NAME_PROPERTY = "CacheInvalidator.ActiveMQ.ProducerName";
+
+    // Cache name prefix of local cache.
+    public static final String LOCAL_CACHE_PREFIX = "$__local__$.";
+
+    // Cache name prefix of clear all.
+    public static final String CLEAR_ALL_PREFIX = "$__clear__all__$.";
+
+    // Default producer retry limit.
+    public static final int PRODUCER_RETRY_LIMIT = 30;
```
Comment (inthirakumaaran) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/CacheSyncUtils.java:42]: Better if we can parameterize this one as well

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/ActiveMQConsumer.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            }
+        }
+        return instance;
+    }
+
+    public void startService() {
+
+        if (!CacheSyncUtils.isActiveMQCacheInvalidatorEnabled()) {
+            log.debug("ActiveMQ based cache invalidation is not enabled");
+            // Todo remove this info log.
```
Comment (inthirakumaaran) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/ActiveMQConsumer.java]: Need to remove all the info logs mentioned in the **todo**

**Comment by inthirakumaaran** [comment]
Need to create the Framework PR for `deployment.toml` configurations.

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+   <ActiveMQ>
+   <BrokerURL>{{cache_invalidator.active_mq.broker_url}}</BrokerURL>
+   <Enabled>{{cache_invalidator.active_mq.enabled}}</Enabled>
+   <TopicName>{{cache_invalidator.active_mq.topic_name}}</TopicName>
+   <ProducerName>{{cache_invalidator.active_mq.producer_name}}</ProducerName>
+   </ActiveMQ>
+   </CacheInvalidator>
+   {% endif %}
+```
+3. Then add the following config
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/README.md]:
```suggestion
3. Then add the following config in the `deployment.toml` file
```

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/pom.xml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            <id>wso2-nexus</id>
+            <name>WSO2 internal Repository</name>
+            <url>http://maven.wso2.org/nexus/content/groups/wso2-public/</url>
+            <releases>
+                <updatePolicy>daily</updatePolicy>
+                <checksumPolicy>ignore</checksumPolicy>
+            </releases>
+        </repository>
+    </repositories>
+
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/pom.xml:73]: unnecessary new line

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerServiceComponent.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * Service component for the ActiveMQ cache manager.
+ */
+@Component(
+        name = "org.wso2.carbon.cache.sync.active.mq.manager.CacheSyncActiveMQManagerServiceComponent",
+        immediate = true
+)
+public class CacheSyncActiveMQManagerServiceComponent {
+
+    private static Log log = LogFactory.getLog(CacheSyncActiveMQManagerServiceComponent.class);
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerServiceComponent.java]: can't we have this variable as final

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerServiceComponent.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    private ServiceRegistration serviceRegistration1 = null;
+    private ServiceRegistration serviceRegistration2 = null;
+    private ServiceRegistration serviceRegistration3 = null;
+    private ServiceRegistration serviceRegistration4 = null;
+    private ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerServiceComponent.java:55]: unnecessary new line

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerDataHolder.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+public class CacheSyncActiveMQManagerDataHolder {
+
+    private static CacheSyncActiveMQManagerDataHolder instance = new CacheSyncActiveMQManagerDataHolder();
+
+
+    public static CacheSyncActiveMQManagerDataHolder getInstance() {
+
+        return instance;
+    }
+
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerDataHolder.java]: unnecessary new line

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerDataHolder.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+package org.wso2.carbon.cache.sync.active.mq.manager.internal;
+
+/**
+ * Data holder for the ActiveMQ cache manager.
+ */
+public class CacheSyncActiveMQManagerDataHolder {
+
+    private static CacheSyncActiveMQManagerDataHolder instance = new CacheSyncActiveMQManagerDataHolder();
+
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/internal/CacheSyncActiveMQManagerDataHolder.java:27]: unnecessary new line

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/ActiveMQProducer.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+                TextMessage message = session.createTextMessage(clusterCacheInvalidationRequest.toString());
+                message.setStringProperty("sender", CacheSyncUtils.getProducerName());
+                producer.send(message);
+                log.info("......................sending from producer...................." +
+                        clusterCacheInvalidationRequest.toString());
+
+                // Clean up resources
+                producer.close();
+                session.close();
+                connection.close();
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/ActiveMQProducer.java]: Shall we do the resource cleaning inside a `final` block

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/pom.xml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+                        <Embed-Dependency>
+                            activemq-all
+                        </Embed-Dependency>
+                        <Embed-Transitive>true</Embed-Transitive>
+                        <Export-Package>
+                            !org.wso2.carbon.cache.sync.active.mq.manager.internal,
+                            org.wso2.carbon.cache.sync.active.mq.manager.*,
+                        </Export-Package>
+                        <Import-Package>
+                            javax.cache;version="[1.0,2)",
```
Comment (pamodaaw) [inline components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/pom.xml]: shall we use variables to define the version ranges

**Review by pamodaaw — APPROVED** [review]

File: components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/CacheSyncUtils.java (hunk 2 above)
Comment (inthirakumaaran) [reply components/active-mq-support/org.wso2.carbon.cache.sync.active.mq/src/main/java/org/wso2/carbon/cache/sync/active/mq/manager/CacheSyncUtils.java:42]: Not needed, retry will be handled from the ActiveMQ side(using the failover concept.)

**Pull request merged**